apps/emqx_durable_storage/src/emqx_ds_beamformer_rt.erl

```diff
@@ -101,46 +120,42 @@ init([CBM, DBShard, Name, _Opts]) ->
         sub_tab = emqx_ds_beamformer:make_subtab(DBShard),
         metrics_id = emqx_ds_beamformer:metrics_id(DBShard, rt),
         name = Name,
-        high_watermark = ets:new(high_watermark, [ordered_set, private]),
+        high_watermark = ets:new(high_watermark, [ordered_set, public]),
         queue = emqx_ds_beamformer_waitq:new(),
         batch_size = emqx_ds_beamformer:cfg_batch_size()
     },
     self() ! ?housekeeping_loop,
     {ok, S}.
 
-handle_call(Req = #sub_state{}, _From, S) ->
-    Reply = do_enqueue(Req, S),
-    {reply, Reply, S};
+handle_call(Req = #sub_state{}, From, S0) ->
+    Fun = fun(S) ->
+        do_enqueue(Req, From, S)
+    end,
+    push_cmd(S0, Fun);
 handle_call(_Call, _From, S) ->
     {reply, {error, unknown_call}, S}.
 
-handle_cast(#seal_event{rank = Rank}, S = #s{queue = Queue}) ->
-    %% Currently generation seal events are treated as stream events,
-    %% for each known stream that has the matching rank:
-    Streams = emqx_ds_beamformer_waitq:streams_of_rank(Rank, Queue),
-    _ = [process_stream_event(false, Stream, S) || Stream <- Streams],
-    {noreply, S};
+handle_cast(#seal_event{rank = Rank}, S0) ->
+    Fun = fun(S = #s{queue = Queue}) ->
+        %% Currently generation seal events are treated as stream events,
+        %% for each known stream that has the matching rank:
+        Streams = emqx_ds_beamformer_waitq:streams_of_rank(Rank, Queue),
+        %% TODO: pass it through the event queue too
+        _ = [process_stream_event(self(), false, Stream, S) || Stream <- Streams],
+        S
+    end,
+    push_cmd(S0, Fun);
 handle_cast(_Cast, S) ->
     {noreply, S}.
 
-handle_info(E = #shard_event{event = Event}, S = #s{shard = Shard, queue = Queue}) ->
-    ?tp(info, beamformer_rt_event, #{event => Event, shard => Shard}),
-    %% Before processing an event, clear the mailbox from the matching
-    %% events to avoid repeatedly hammering the DB:
-    flush_similar_events(E),
-    %% TODO: make a proper event wrapper?
-    Stream = Event,
-    _ =
-        case emqx_ds_beamformer_waitq:has_candidates(Stream, Queue) of
-            true ->
-                process_stream_event(true, Stream, S);
-            false ->
-                %% Even if we don't have any poll requests for the stream,
-                %% it's still necessary to update the high watermark to
-                %% avoid full scans of very old data:
-                update_high_watermark(Stream, S)
-        end,
-    {noreply, S};
+handle_info(E = #shard_event{}, S = #s{shard = Shard, active_streams = Q}) ->
+    ?tp(info, beamformer_rt_event, #{event => E, shard => Shard}),
```

Comment:
Nit. Volume of such events might be too high for `info` level?

Reply:
True, it was a forgotten debug message. Changed the severity to debug.